proto/pulumi/provider.proto

```diff
@@ -215,32 +220,105 @@ message GetSchemaResponse {
+    // whether or not the provider accepts secrets in general, providers *must* handle secrets if they appear in `args`.
+    // :::
+    google.protobuf.Struct args = 2;
+
+    // True if and only if the caller supports secrets. If true, operations should return strongly typed secrets if the
+    // provider supports them also.
+    bool acceptSecrets  = 3;
+
+    // True if and only if the caller supports strongly typed resources. If true, operations should return resources as
+    // strongly typed values if the provider supports them also.
+    bool acceptResources = 4;
+
+    // True if and only if the caller supports sending old inputs as part of [](pulumirpc.ResourceProvider.Diff) and
+    // [](pulumirpc.ResourceProvider.Update) calls. If true, the provider should expect these fields to be populated in
+    // these calls.
+    bool sends_old_inputs = 5;
+
+    // True if and only if the caller supports sending old inputs and outputs as part of
+    // [](pulumirpc.ResourceProvider.Delete) calls. If true, the provider should expect these fields to be populated in
+    // these calls.
+    bool sends_old_inputs_to_delete = 6;
 }
 
+// `ConfigureResponse` is the type of responses sent by a [](pulumirpc.ResourceProvider.Configure) call. Its primary
+// purpose is to communicate features that the provider supports back to the caller.
 message ConfigureResponse {
-    bool acceptSecrets = 1;   // when true, the engine should pass secrets as strongly typed values to the provider.
-    bool supportsPreview = 2; // when true, the engine should invoke create and update with preview=true during previews.
-    bool acceptResources = 3; // when true, the engine should pass resources as strongly typed values to the provider.
-    bool acceptOutputs = 4;   // when true, the engine should pass output values to the provider.
+    // True if and only if the provider supports secrets. If true, the caller should pass secrets as strongly typed
+    // values to the provider.
+    bool acceptSecrets = 1;
+
+    // True if and only if the provider supports the `preview` field on [](pulumirpc.ResourceProvider.Create) and
+    // [](pulumirpc.ResourceProvider.Update) calls. If true, the engine should invoke these calls with `preview` set to
+    // `true` during previews.
+    bool supportsPreview = 2;
+
+    // True if and only if the provider supports strongly typed resources. If true, the caller should pass resources as
+    // strongly typed values to the provider.
+    bool acceptResources = 3;
+
+    // True if and only if the provider supports output values as inputs. If true, the engine should pass output values
+    // to the provider where possible.
+    bool acceptOutputs = 4;
 }
 
-// ConfigureErrorMissingKeys is sent as a Detail on an error returned from `ResourceProvider.Configure`.
+// `ConfigureErrorMissingKeys` is the type of error details that may be sent in response to a
```

Comment:
So error handling approach might need to be documented. There are two cases of errors, problems with user input that the user might need to take action on and problems with the system.

The common error highlighted here is that required configuration keys are missing.

Another common error is that the provider does not recognize a configuration property the user sent, perhaps because it is mis-spelled, and wants to tell the user to fix their program.

Another common error is that the provider is done with semantics checks and the configuration is reasonable but the provider attempts to verify that it can connect to its cloud with the credentials supplied in the configuration. If it cannot do so it wants to communicate to the user a message also indicating what is wrong. 

These considerations  may not belong here but in a provider guide or dedicated pages, but the questions are which mechanism is preferred for these errors: 

1. emitting from CheckConfig or Configure
2. emitting through gRPC special errors or response fields
3. how to "officially" tie errors to individual configuration properties, which can be nested